src/antidote_locks.md

```diff
@@ -0,0 +1,120 @@
+2. Check if we have all required locks.
+    - If a lock is missing, send a request to all other DCs and wait for responses.
+3. Once all locks are locally available: 
+    - Wait until there are no other known requests with conflicting locks that should be served first (or already hold a lock locally)
+    - Acquire the locks and reply with the current snapshot time.
+
+
+When receiving a lock request from a different DC:
+
+- For each requested lock:
+    1. Wait until there are no other known requests with conflicting locks that should be served first (or already hold a lock locally).
+    2. Update the CRDT state of the locks to transfer.
+    3. Send an inter-DC message to the requesting DC.
+
+### Ordering:
+
+To guarantee liveness and fairness, we use the following ordering on requests:
+
+- Each lock request receives a timestamp that is used for ordering.
+    - If all locks are available locally, this is the current system time.
+    - Otherwise, it is the current system time plus the value of `INTER_DC_LOCK_REQUEST_DELAY`.
+        This is the expected time we have to wait to get locks from other DCs, so we add this time such that local requests can still use the locks for some time.
+- Lock requests are ordered lexicographically by the pair `{RequestTime, RequestingDc}` (i.e. the DC identifier is used to arbitrate in case of equal request times).
+
+### Performance
+
+Since it is quite expensive to move locks from one DC to another (the other DC has to catch up to the same snapshot time), we want to avoid sending a lock to a different DC if it is still required locally.
+However, it should eventually be sent to the other DC to guarantee fairness.
+
+To achieve this goal, we use the following strategy:
+
+Always send the lock to the requesting DC if it has been held locally for more than `?MAX_LOCK_HOLD_DURATION`.
+Otherwise, only send the lock if it has not been used in the last `?MIN_EXCLUSIVE_LOCK_DURATION` (since it is likely to be used again).
+ 
+
+
+
+### Code Structure
+
+- `clocksi_interactive_coord.erl`:
+    - Locks are acquired in `start_tx_internal`.
+    - Locks are released in `before_commit_checks`.
+- `antidote_locks.erl`: 
+    - Provides the internal lock API
+- `antidote_lock_server`:
+    - A gen_server with one instance per DC (the server is registered using the global name `antidote_lock_server`).
+    - Responsible for handling the requests.
+- `antidote_lock_server_state`:
+    - Encapsulates the state and logic of the `antidote_lock_server`.
+    - This module is purely functional such that it is easier to test.
+- `antidote_lock_server_sup`:
+    - Supervisor for the `antidote_lock_server`
+    
+
+
+
+Future Work
+-----------
+
+- *Dynamic Membership*: Adding and removing DCs is currently not supported.
+- *Fault tolerance*: In the following scenarios, the current implementation will not work correctly:
+    - CRDT state cannot be updated locally: Implementation will wait for the update indefinitely.
```

Comment:
I don't understand this comment: Which CRDT state update is blocking here?

Reply:
When a lock is updated I am using an CRDT update operation. This update might fail, e.g. if the partition with the lock is currently not available. There is no error handling for this case yet.